raft/fsm.cc

```diff
@@ -1019,8 +1019,12 @@ bool fsm::apply_snapshot(snapshot_descriptor snp, size_t max_trailing_entries, s
     // If the snapshot is local, _commit_idx is larger than snp.idx.
     // Otherwise snp.idx becomes the new commit index.
     _commit_idx = std::max(_commit_idx, snp.idx);
-    _output.snp.emplace(fsm_output::applied_snapshot{snp, local, max_trailing_entries});
-    size_t units = _log.apply_snapshot(std::move(snp), max_trailing_entries, max_trailing_bytes);
+    auto [units, new_first_index] =
+        _log.apply_snapshot(std::move(snp), max_trailing_entries, max_trailing_bytes);
+    _output.snp.emplace(
+        fsm_output::applied_snapshot{.snp = _log.get_snapshot(),
+                                     .is_local = local,
+                                     .preserve_log_entries = _log.get_snapshot().idx - new_first_index});
```

Comment:
Why not return `new_first_index` instead? With this approach, inside raft_sys_table_storage.cc we have to calculate first index again, and although it's a simple calculation:
```cpp
future<> raft_sys_table_storage::update_snapshot_and_truncate_log_tail(const raft::snapshot_descriptor &snap, size_t preserve_log_entries) {
    // Update snapshot and truncate logs in `system.raft` atomically
    raft::index_t log_tail_idx = raft::index_t(static_cast<uint64_t>(snap.idx) - static_cast<uint64_t>(preserve_log_entries));
```

it could still be considered duplication of logic, preferably to be avoided.

I don't have a strong opinion about this however.

Reply:
I thought about that in the beginning, but then found that it was the original API of that function - to take the snapshot and the number of  items that need to be preserved prior to the snapshot index, and saw some tests that use that API, so decided to leave it as it is.  Let me take another look...

Reply:
If this is a "public API" (in particular used by tests), API backward compatibility is a good argument.